### ISSUE TITLE
Training with COCO dataset and additional classes

### ISSUE THREAD

**Issue by kenrubiooo** [AlexeyAB/darknet#2147]
Hi @AlexeyAB,

I have questions on training using the COCO dataset with addition of images for new classes. I want to train the YOLOv3 algorithm on 80 classes from COCO dataset + 7 more classes from my custom dataset (2,000 images each class). These 14,000 images are single objects in an image, so 1 object in 1 image.

Questions:
1) Do I need to recalculate the anchors?

2) Do I need to train the algorithm with (i) all the images from the 87 classes or  (ii) only the images from the additional classes? If it is (ii), how do I do it?

3) How many iterations should I do this and where to set the max iterations? Do I set is in 

> max_batches

?

4) How do I start training? Do I set param

>  stopbackward=1

 here: (https://github.com/AlexeyAB/darknet/blob/6d44529cf93211c319813c90e0c1adb34426abe5/cfg/yolov3.cfg#L548) 

then do this command: 

> ./darknet partial cfg/yolov3.cfg yolov3.weights yolov3.conv.81 81

, then train by using weights file yolov3.conv.81? 

Or do I need to use the 

> darknet53.conv.74 for training?

Thank you!

**Comment by AlexeyAB** [comment]
@kenrubiooo Hi,

1. It isn't necessary. But you can.

2. You must train by using all images (COCO + your dataset) with all 87 classes

3. You should train at least several epochs, f.e. `4 epochs = total_images / 64` iterations. So how many images do you have in total? Train the model until mAP will stop increase: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

4. You should do, and don't set `stopbackward`
```
./darknet partial cfg/yolov3.cfg yolov3.weights yolov3.conv.81
./darknet detector train cfg/coco_new.data cfg/yolov3_new.cfg yolov3.conv.81 -map
```

**Comment by kenrubiooo** [comment]
Thank you, @AlexeyAB!

3. Since COCO2014 has 80k training images, I have approximately 94,000 images in total, that is 1,469 epochs = 94,000 / 64.

4a. Why do I not need to use the stopbackward? 

4b. Also, why is `yolov3.cfg` in the first command and `yolov3_new.cfg` is used in the second command? What is with the yolov3.cfg that is not found in the yolov3_new.cfg?

**Comment by AlexeyAB** [comment]
3. Yes, train the model at least 6 000 iterations or more. (while 1 epoch = 1500 iterations). And look at the mAP chart since you train with `-map` flag.

4a You need not to use stopbackward. But if you want you can use it here: https://github.com/AlexeyAB/darknet/blob/48d461f9bda2d01dc2a83c7c2a520f61b3ea5b79/cfg/yolov3.cfg#L548

More: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> to speedup training (with decreasing detection accuracy) do Fine-Tuning instead of Transfer-Learning, set param stopbackward=1 here: https://github.com/AlexeyAB/darknet/blob/6d44529cf93211c319813c90e0c1adb34426abe5/cfg/yolov3.cfg#L548 then do this command: ./darknet partial cfg/yolov3.cfg yolov3.weights yolov3.conv.81 81 will be created file yolov3.conv.81, then train by using weights file yolov3.conv.81 instead of darknet53.conv.74

4b You can use `yolov3_new.cfg` in both commands.

**Comment by kenrubiooo** [comment]
Thank you again, @AlexeyAB!

I somehow understand the fine-tuning and transfer learning difference. 

Can you clarify the significance of the first command:
`./darknet partial cfg/yolov3.cfg yolov3.weights yolov3.conv.81`

What is in yolov3.cfg that you specified it here instead of the yolov3_new.cfg? Assume that I changed the batch size, subdivision, classes, filters, etc. Also, assume that the yolov3.cfg was not changed from your repository.

Thank you!

**Comment by kenrubiooo** [comment]
Sorry for the late reply, @AlexeyAB. Was out the whole day.

Thank you for the explanation. So, it is just freezing the first 81 layers. 

Next questions, sorry. 

1) In training with multiple GPUs, do I still use the commands 
```
./darknet partial cfg/yolov3.cfg yolov3.weights yolov3.conv.81
./darknet detector train cfg/coco_new.data cfg/yolov3_new.cfg yolov3.conv.81 -map
```
to start the training?

2) Why do I need to train the model for about 1000 iterations with only 1 GPU then use the partially trained model weights to continue training with multiple GPUs?

3) How does training a detector model in multiple GPUs work? How do multiple GPUs create a single weights file since the processes are cut into pieces?

**Comment by AlexeyAB** [comment]
Use this command for training using multiple GPUs: 
```
./darknet partial cfg/yolov3.cfg yolov3.weights yolov3.conv.81
./darknet detector train cfg/coco_new.data cfg/yolov3_new.cfg yolov3.conv.81 -map
./darknet detector train cfg/coco_new.data cfg/yolov3_new.cfg backup/yolov3_new_1000.weights -map -gpus 0,1,2,3
```

More about it: https://github.com/AlexeyAB/darknet#how-to-train-with-multi-gpu

**Comment by kenrubiooo** [comment]
Hi @AlexeyAB!

I followed this set of codes for multiple gpus. After 1,000 iterations, I used 4 gpus in total. What happened is that the average loss changed from 4.6 to 5,000+ average error.

**Comment by AlexeyAB** [comment]
@kenrubiooo Hi,
Somethis went wrong.
Try to train about 4000 iterations with 1 GPU and then switch to 4 GPUs.
And check your cfg-file.

https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> Note: **If during training you see nan values for avg (loss) field - then training goes wrong**, but if nan is in some other lines - then training goes well.

**Comment by kenrubiooo** [comment]
I tried to train for about 7000 since I left my laptop open for a while. Then I tried to train on the server using 4 GPUs. Everything seems to be working well right now. 

Maybe the reason why the loss jumped is because of the instability? I read in the groups that multi-gpu training is not yet stable.

**Comment by kenrubiooo** [comment]
Hi @AlexeyAB,

I am still training the model. It is now at 13,464 iterations. I have not seen any values of mAP on the screen. I have 96,801 images in total. 

On the terminal, I am seeing 
`calculation mAP (mean average precision)...`
and some number increasing.

**Comment by kenrubiooo** [comment]
Sorry, I just needed for it to finish. I have now seen one map after 13,464 iterations. Currently, it is 48%. Thank you very much for the help! I will ask whenever something comes up.

**Comment by kenrubiooo** [comment]
Hi @AlexeyAB,

Does this repo have a timer for the training? Start time and end time whenever there is early stopping, not necessarily when the whole training finishes (reached final iteration). I am trying to time the training whenever there is additional number of classes. Thank you!

**Comment by AlexeyAB** [comment]
@kenrubiooo Hi,
No, there is no timer.

**Comment by kenrubiooo** [comment]
Ok, @AlexeyAB. Thank you! I had a mAP of 54% and tried it out for inferencing.

So, what happened was I cannot detect any object from the new classes most of the time. When I do , sometimes the bounding boxes are too big. Most of the time, I am misclassified as a shoe, for example. I think this is because of the training images of the new classes I am adding. I only have resized and rotated images of the original images. All of the images for the new classes have the same background, i.e. white. I think the white background is causing this error since the background is not generalized. What do you think?

**Comment by AlexeyAB** [comment]
@kenrubiooo Hi,

* How many iterations did you train?

* What `AP`s can you get for the new 7 classes? 

* Can you show output of `./darknet detector map ` command with: AP, TP/FP/FN, mAP?

----

> + 7 more classes from my custom dataset

Can you attach 1-2 txt-files with annotation for the new classes?

----

> I only have resized and rotated images of the original images. 

Can you show examples?

----

>  All of the images for the new classes have the same background, i.e. white.

Yes, this is not the best images for training. You should try use Training images that are similar to Test images.

----

> So, what happened was I cannot detect any object from the new classes most of the time. 

Can you show an example?
You can't detect any object even on the Training images?

**Comment by AlexeyAB** [comment]
@kenrubiooo 

> When I use a webcam, all the detections have misclassifications.

Can you show objects from the Web-cam that you can't detect?

Check that for each object which you want to detect - there is at least 1 similar object in the Training dataset with about the same: shape, side of object, relative size, angle of rotation, tilt, illumination.

Check whether you really want to detect sneakers that are upside down on the image.

https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> for each object which you want to detect - there must be at least 1 similar object in the Training dataset with about the same: shape, side of object, relative size, angle of rotation, tilt, illumination. So desirable that your training dataset include images with objects at diffrent: scales, rotations, lightings, from different sides, on different backgrounds - you should preferably have 2000 different images for each class or more, and you should train 2000*classes iterations or more
